src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+struct dir_obj {
+	int              fd;
+	uint32_t         num_ents;
+	dfs_obj_t       *dir;
+	long int         offset;
+	struct dfs_mt   *dfs_mt;
+	int              open_flag;
+	int              ref_count;
+	unsigned int     st_ino;
+	daos_anchor_t    anchor;
+	/* path and ents will be allocated together dynamically since they are large. */
+	char            *path;
+	struct dirent   *ents;
+};
+
+struct mmap_obj {
+	/* The base address of this memory block */
+	char            *addr;
+	size_t           length;
+	/* the size of file. It is needed when write back to storage. */
+	size_t           file_size;
+	int              prot;
+	int              flags;
+	/* The fd used when mmap is called */
+	int              fd;
+	/* num_pages = length / page_size */
+	int              num_pages;
+	int              num_dirty_pages;
+	off_t            offset;
+	/* An array to indicate whether a page is updated or not */
+	bool            *updated;
+};
+
+struct fd_dup2 {
+	int fd_src, fd_dest;
+	bool dest_closed;
+};
+
+/* working dir of current process */
+static char            cur_dir[DFS_MAX_PATH] = "";
+static bool            segv_handler_inited;
+/* Old segv handler */
+struct sigaction       old_segv;
+
+/* the flag to indicate whether initlization is finished or not */
+static bool            hook_enabled;
+static pthread_mutex_t lock_fd;
+static pthread_mutex_t lock_dirfd;
+static pthread_mutex_t lock_mmap;
+
+/* store ! umask to apply on mode when creating file to honor system umask */
+static mode_t          mode_not_umask;
+
+static int             fd_stdin = -1, fd_stdout = -1, fd_stderr = -1;
+
+static void
+finalize_dfs(void);
+static void
+update_cwd(void);
+
+struct dir_hdl {
+	d_list_t   entry;
+	dfs_obj_t *oh;
+	char       name[DFS_MAX_PATH];
+};
+
+static inline struct dir_hdl *
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dir_hdl, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return (strcmp(hdl->name, (const char *)key) == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int             rc;
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	rc = dfs_release(hdl->oh);
+	if (rc == ENOMEM)
+		rc = dfs_release(hdl->oh);
+	if (rc) {
+		printf("Error: dfs_release() failed: (%s)", strerror(rc));
+		exit(1);
+	}
+	free(hdl);
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	return true;
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->name, strlen(hdl->name));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
+					  .hop_rec_decref = rec_decref,
+					  .hop_rec_free   = rec_free,
+					  .hop_rec_hash   = rec_hash};
+
+static dfs_obj_t *
+lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode)
```

Comment:
This function is not thread safe and doesn't appear to be protected by locks, but I notice other parts of the code does have locking.

Reply:
Both d_hash_rec_find() and d_hash_rec_insert() have locks since D_HASH_FT_MUTEX is used in d_hash_table_create(). Do we need more locks inside lookup_insert_dir()?